core/embed/rust/src/ui/display.rs

```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+/// and optionally draw an `icon` inside
+pub fn rect_rounded2_partial(
+    area: Rect,
+    fg_color: Color,
+    bg_color: Color,
+    show_percent: i32,
+    icon: Option<(&[u8], Color)>,
+) {
+    const MAX_ICON_SIZE: u16 = 64;
+
+    let r = area.translate(get_offset());
+    let clamped = r.clamp(constant::screen());
+
+    set_window(clamped);
+
+    let center = r.center();
+    let colortable = get_color_table(fg_color, bg_color);
+    let mut icon_colortable = colortable;
+
+    let mut use_icon = false;
+    let mut icon_area = Rect::zero();
+    let mut icon_area_clamped = Rect::zero();
+    let mut icon_data = [0_u8; ((MAX_ICON_SIZE * MAX_ICON_SIZE) / 2) as usize];
+    let mut icon_width = 0;
+
+    if let Some(i) = icon {
+        let toif_info = display::toif_info(i.0).unwrap();
+        assert!(toif_info.grayscale);
+
+        if toif_info.width <= MAX_ICON_SIZE && toif_info.height <= MAX_ICON_SIZE {
+            icon_area = Rect::from_center_and_size(
+                center,
+                Offset::new(toif_info.width.into(), toif_info.height.into()),
+            );
+            icon_area_clamped = icon_area.clamp(constant::screen());
+
+            let mut ctx = uzlib::UzlibContext::new(&i.0[12..], false);
+            if ctx.uncompress(&mut icon_data).is_err() {
+                return;
+            }
+            icon_colortable = get_color_table(i.1, bg_color);
+            icon_width = toif_info.width.into();
+            use_icon = true;
+        }
+    }
+
+    let start = 0;
+    let end = (start + ((360 * show_percent) / 100)) % 360;
+
+    let start_vector;
+    let end_vector;
+
+    let mut show_all = false;
+    let mut inverted = false;
+
+    if show_percent >= 100 {
+        show_all = true;
+        start_vector = Point::zero();
+        end_vector = Point::zero();
+    } else if show_percent > 50 {
+        inverted = true;
+        start_vector = get_vector(end);
+        end_vector = get_vector(start);
+    } else {
+        start_vector = get_vector(start);
+        end_vector = get_vector(end);
+    }
+
+    let n_start = Point::new(-start_vector.y, start_vector.x);
+
+    for y_c in r.y0..r.y1 {
+        for x_c in r.x0..r.x1 {
+            let p = Point::new(x_c, y_c);
+
+            let mut icon_pixel = false;
+            if use_icon && icon_area_clamped.contains(p) {
+                let x_i = p.x - icon_area.x0;
+                let y_i = p.y - icon_area.y0;
+
+                let data = icon_data[(((x_i & 0xFE) + (y_i * icon_width)) / 2) as usize];
+                if (x_i & 0x01) == 0 {
+                    pixeldata(icon_colortable[(data >> 4) as usize]);
+                } else {
+                    pixeldata(icon_colortable[(data & 0xF) as usize]);
+                }
+                icon_pixel = true;
+            }
+
+            if !clamped.contains(p) || icon_pixel {
+                continue;
+            }
+
+            if !icon_pixel {
```

Comment:
it seems icon_pixel must be false here because of the above condition?

Reply:
true: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2